r/R/util.R

```diff
@@ -217,8 +225,27 @@ handle_csv_read_error <- function(e, schema, call) {
         "header being read in as data."
       )
     )
+    abort(msg, call = call)
+  }
+}
+
+# This function only raises an error if
+# the appropriate string was found and so errors must be raised manually after
+# calling this if matching error not found
+# TODO: Refactor as part of ARROW-17355 to prevent potential missed errors
+handle_augmented_field_misuse <- function(e, call) {
+  msg <- conditionMessage(e)
+  if (grepl("No match for FieldRef.Name(__filename)", msg, fixed = TRUE)) {
+    msg <- c(
+      msg,
+      i = paste(
+        "Augmented fields such as 'filename' must",
+        "only be used with with Dataset objects which have",
+        "not been aggregated or joined."
```

Comment:
Wordsmithing here, how about something like "'filename' can only be used with Dataset objects, and it can only be added before doing an aggregation or a join"?

Reply:
That's a lot clearer, thank you!  I've updated the error to incorporate both that, and the two different ways of referring to the `__filename` variable.